Match in match heuristic

bclement-ocp opened the pull request.

bclement-ocp committed:
Add match-in-match heuristic + tests

Co-authored-by: Nathanaëlle Courant <Ekdohibs@users.noreply.github.com>

Ekdohibs commented on middle_end/flambda2/simplify/simplify_switch_expr.ml:
```diff
@@ -667,6 +667,169 @@ let simplify_arm ~typing_env_at_use ~scrutinee_ty arm action (arms, dacc) =
+          | `Spec (join_analysis, specialized, generic) ->
+            (* Specialization benefit estimation: we use heuristics similar to
+               that of inlining to estimate the benefit based on code size and
+               removed operations (note that we use the join info the the typing
```
```suggestion
               removed operations (note that we use the join info in the typing
```
?

Ekdohibs reviewed (commented):
There are still 6 example files at toplevel, could we move them to the testsuite? (maybe checking the fexpr output, but at least just checking they compile correctly).

Gbury committed:
fix test

Ekdohibs commented on driver/oxcaml_flags.ml:
```diff
@@ -230,7 +233,7 @@ module Flambda2 = struct
   let o3 = {
     o2 with
-    function_result_types = Functors_only
+    function_result_types = Functors_only;
```
```suggestion
    function_result_types = Functors_only
```
Stray change?

Gbury replied:
No idea, I'd guess it's an ocamlformat thing ?

Gbury replied:
So apparently, ocamlformat keeps the original formatting for this very particular change, because reasons I guess ?

Gbury committed:
Apply suggestion from @Ekdohibs

Co-authored-by: Nathanaëlle Courant <Ekdohibs@users.noreply.github.com>

bclement-ocp commented on driver/oxcaml_flags.ml:
```diff
@@ -328,9 +332,16 @@ module Flambda2 = struct
+      cont_spec_threshold = 0.;
+    }
 
     let o4 = default
```
This should be `let o4 = o3` I think.

Gbury replied:
done (as well as slightly raising the lifting budget for O4)

Gbury committed:
Add match-in-match to O4

Gbury committed:
Fix runtest-upstream

Gbury committed:
fix dune tests

Gbury committed:
even better options for o4

Gbury committed:
deactivate heuristice in o4, and fix options in CI

Gbury committed:
fix Ci options

Gbury committed:
make fmt

Gbury committed:
update tests

Gbury committed:
Disable failing test

Match-in-match changes the ouput of the test, while still being correct,
but it's a pain to try and have different output depending on the
configuration of the CI.

Gbury committed:
let's try ci again

mshinwell committed:
Merge remote-tracking branch 'oxcaml/main' into match_in_match_heuristic

mshinwell committed:
Typo
